rust/psibase_macros/psibase-macros-lib/src/tests/basics.rs

```diff
@@ -0,0 +1,72 @@
+#[cfg(test)]
+mod tests {
+    use crate::service_macro::service_macro_impl;
+
+    #[test]
+    #[should_panic]
+    // Events can only have one of 3 values: history, ui, or merkle
+    // TODO: How to test the error text?
+    fn test_event_attribute_parsing_event_type() {
+        use quote::quote;
+
+        let attr = quote! {};
+        let before = quote! {
+            mod service {
+                #[action]
+                fn add(a: i32, b: i32) -> i32 {
+                    a + b
+                }
+                #[event(historyui)]
+                fn something() {
+
+                }
+            }
+        };
+
+        let _after = service_macro_impl(attr, before);
+    }
+
+    #[test]
+    #[should_panic]
+    // Events can only have one event type
+    fn test_event_attribute_parsing_single_event_type() {
+        use quote::quote;
+
+        let attr = quote! {};
+        let before = quote! {
+            mod service {
+                #[action]
+                fn add(a: i32, b: i32) -> i32 {
+                    a + b
+                }
+                #[event(history, ui)]
+                fn something() {
+
+                }
+            }
+        };
+
+        let _after = service_macro_impl(attr, before);
+    }
+
+    #[test]
+    // `service` macro should expand without requiring a `use` to provide its deps
+    fn test_macro_hygiene() {
+        use quote::quote;
+
+        let attr = quote! {};
+        let before = quote! {
+            mod service {
+
+                #[action]
+                fn add(a: i32, b: i32) -> i32 {
+                    a + b
+                }
+                #[event(history)]
+                fn something() {}
+            }
+        };
+
+        service_macro_impl(attr, before);
```

Comment:
> `service` macro should expand without requiring a `use` to provide its deps

Does this actually check that the generated TokenStream compiles though?

Reply:
No, it only tests that the macro doesn't find syntax it doesn't recognize.